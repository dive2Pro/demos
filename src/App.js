import React from "react";
import { render } from "react-dom";
// console.clear()
const styles = {
  fontFamily: "sans-serif",
  textAlign: "center"
};
const Modes = {
  Locked: "Locked",
  Home: "Home",
  Unlocked: "Unlocked",
  Desktop: "Desktop",
  App: "App"
};
const ModeMessages = {
  [Modes.Home]: "The phone is at home",
  [Modes.Locked]: "The Phone is Locked",
  [Modes.Unlocked]: "The Phone is Unlocked",
  [Modes.App]: "The Phone is opening App",
  [Modes.Desktop]: "The Phone is switching Desktop"
};

const NotAllowed = {
  mode: "Null",
  message: "The operation is not allowed"
}

class Main extends React.PureComponent {
  state = {
    mode: Modes.Locked,
    message: ModeMessages[Modes.Locked]
  };
  onHomeClick = () => {
    const { mode } = this.state;

    if (mode === Modes.Locked) {
      this.setState({
        mode: Modes.Unlocked,
        message: ModeMessages[Modes.Unlocked]
      });
    } else if (mode === Modes.Unlocked) {
      this.setState({
        mode: Modes.Home,
        message: ModeMessages[Modes.Home]
      });
    } else if (mode === Modes.Desktop) {
      this.setState({
        mode: Modes.Home,
        message: ModeMessages[Modes.Home]
      });
    } else if (mode === Modes.App) {
      this.setState({
        mode: Modes.Home,
        message: ModeMessages[Modes.Home]
      });
    } else {
      this.setState({
        mode: Modes.Home,
        message: ModeMessages[Modes.Home]
      });
    }
  };
  onOpenAppClick = () => {
    const {mode} = this.state
    if (mode === Modes.Desktop || mode === Modes.Home) {
      this.setState({
        mode: Modes.App,
        message: ModeMessages[Modes.App]
      })
    } else {
      this.setState(NotAllowed)
    }
  };
  onSwitchDesktopClick = () => {
    const { mode } = this.state
    
    if (mode === Modes.Home) {
      this.setState ( {
        mode: Modes.Desktop,
        message: ModeMessages[Modes.Desktop]
      })
    } else {
      this.setState(NotAllowed)
      
    }
  };
  _handleClickEvent = e => {
    const { id } = e.target;
    if (id) {
      this[`on${id}Click`]();
    }
  };
  render() {
    return (
      <div onClick={this._handleClickEvent}>
        <button id="Home" data-testid="home">Home</button>
        <button id="OpenApp" data-testid='open-app'>Open App</button>
        <button id="SwitchDesktop" data-testid="switch-desktop" >Switch Desktop</button>
        <h2 data-testid="message">{this.state.message}</h2>
      </div>
    );
  }
}

export const App = () => (
  <div style={styles}>
    <h2>Start here {"\u2728"}</h2>
    <Main />
  </div>
);

export default App