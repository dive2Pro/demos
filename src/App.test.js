import React from "react";
import { render, Simulate, wait } from "react-testing-library";
import { App } from "./App";

// require("react-testing-library/extend-expect")

function innerTextEquals(getByTestId) {
  return msg => {
    const dom = getByTestId("message");
    expect(dom).toMatchSnapshot()
  };
}

describe("every functionality works", () => {
 

  const click = {
    home() {
      Simulate.click(getByTestId('home'))
    },
    switch() {
      Simulate.click(getByTestId('switch-desktop'))
    },
    openApp() {
      Simulate.click(getByTestId('open-app'))
    }
  }
  const backToHome = () => {
    click.home()
    click.home()
  }

  let getByText, getByTestId, container, queryByTestId, unmount;
  let textEquals;
  beforeEach(() => {
    ({ getByText, getByTestId, container, queryByTestId, unmount } = render(
      <App />
    ));
    textEquals = innerTextEquals(getByTestId);
  });
  afterEach(() => {
    if (unmount) {
      unmount();
    }
  });
  it("works", () => {
    expect(container).toBeTruthy();
  });
  it(`默认 state 应该是 Locked`, () => {
    textEquals("The Phone is Locked");
  });

  it(`第一次按下 Home 键, 应该是 Unlocked`, async () => {
    click.home()
    textEquals("The Phone is Unlocked");
  });
  it('第二次按下home, 回到 home', async () => {
    backToHome()
    textEquals("The Phone is at Home");
  })

  it(`当在 home 页面时, openApp 会打开一个 app`, () => {
    backToHome()
    click.openApp()
    textEquals()
  })

  it(`如果不在 Home 页面时,  点击 opening app 会出现错误页面`, () => {
    click.openApp()
    textEquals()
    Simulate.click(getByTestId('home'))
    Simulate.click(getByTestId('open-app'))
    
    // textEquals()
  })

  it(`如果在错误页面只有 返回 home, 其他操作都是不允许的`, () => {
    click.openApp()
    textEquals()

    click.switch()
    textEquals()

    click.openApp()
    textEquals()
  })

  it('在 home 页面, 可以点击 switch-desktop', () => {
    backToHome()
    click.switch()
    textEquals()
  })
  
  it(`在其他页面. 点击 home 都会返回 home 页面`, () => {
    backToHome()
    click.openApp()
    textEquals()
    click.home()
    textEquals()
    click.switch()
    textEquals()
    click.home()
    textEquals()

  })
});
