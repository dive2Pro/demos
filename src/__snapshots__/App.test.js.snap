// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`every functionality works 在 home 页面, 可以点击 switch-desktop 1`] = `
<h2
  data-testid="message"
>
  The Phone is switching Desktop
</h2>
`;

exports[`every functionality works 在其他页面. 点击 home 都会返回 home 页面 1`] = `
<h2
  data-testid="message"
>
  The Phone is opening App
</h2>
`;

exports[`every functionality works 在其他页面. 点击 home 都会返回 home 页面 2`] = `
<h2
  data-testid="message"
>
  The phone is at home
</h2>
`;

exports[`every functionality works 在其他页面. 点击 home 都会返回 home 页面 3`] = `
<h2
  data-testid="message"
>
  The Phone is switching Desktop
</h2>
`;

exports[`every functionality works 在其他页面. 点击 home 都会返回 home 页面 4`] = `
<h2
  data-testid="message"
>
  The phone is at home
</h2>
`;

exports[`every functionality works 如果不在 Home 页面时,  点击 opening app 会出现错误页面 1`] = `
<h2
  data-testid="message"
>
  The operation is not allowed
</h2>
`;

exports[`every functionality works 如果在错误页面只有 返回 home, 其他操作都是不允许的 1`] = `
<h2
  data-testid="message"
>
  The operation is not allowed
</h2>
`;

exports[`every functionality works 如果在错误页面只有 返回 home, 其他操作都是不允许的 2`] = `
<h2
  data-testid="message"
>
  The operation is not allowed
</h2>
`;

exports[`every functionality works 如果在错误页面只有 返回 home, 其他操作都是不允许的 3`] = `
<h2
  data-testid="message"
>
  The operation is not allowed
</h2>
`;

exports[`every functionality works 当在 home 页面时, openApp 会打开一个 app 1`] = `
<h2
  data-testid="message"
>
  The Phone is opening App
</h2>
`;

exports[`every functionality works 第一次按下 Home 键, 应该是 Unlocked 1`] = `
<h2
  data-testid="message"
>
  The Phone is Unlocked
</h2>
`;

exports[`every functionality works 第二次按下home, 回到 home 1`] = `
<h2
  data-testid="message"
>
  The phone is at home
</h2>
`;

exports[`every functionality works 默认 state 应该是 Locked 1`] = `
<h2
  data-testid="message"
>
  The Phone is Locked
</h2>
`;
